qualtran/bloqs/multiplexers/apply_lth_bloq.py

```diff
@@ -0,0 +1,100 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Sequence, Tuple
+
+import cirq
+import numpy as np
+from attrs import field, frozen, validators
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, BoundedQUInt, QBit, Register, Side
+from qualtran.bloqs.multiplexers.unary_iteration_bloq import UnaryIterationGate
+
+
+@frozen
+class ApplyLthBloq(UnaryIterationGate):
+    r"""A SELECT operation that executes one of a list of bloqs $U_l$ based on a quantum index:
+
+    $$
+    \mathrm{SELECT} = \sum_{l}|l \rangle \langle l| \otimes U_l
+    $$
+
+    This bloq uses the unary iteration scheme to apply `ops[selection]` controlled on an optional single-bit `control` register.
+
+    Args:
+        ops: List of bloqs to select from, each with identical registers that are all THRU.
+        control: If True, a singly controlled gate is constructed.
+
+    Registers:
+        selection: The index of the bloq in `ops` to execute.
+        control: The control bit if specified above.
+        [user_spec]: The output registers of the bloqs in `ops`.
+
+    References:
+        [Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity](https://arxiv.org/abs/1805.03662).
+        Babbush et. al. (2018). Section III.A. and Figure 7.
+    """
+
+    ops: Tuple[Bloq, ...] = field(
+        converter=lambda x: x if isinstance(x, tuple) else tuple(x), validator=validators.min_len(1)
+    )
```

Comment:
In fact, consider accepting a ndarray instead of a Tuple and make it a `Union[Shaped, NDArray[Bloq]]` so that users can specify N-dimensional SELECT operations.

The SELECT operations are often controlled on multiple selection registers. Expecting a list os `ops` implies that you expect users to create a single "contiguous" register and represent your LCU as a summation of a single variable. However, it's often preferred and is more optimal to keep the summation indices factorized and implement a SELECT operation multiplexed by multiple selection registers. 

For example, see Appendix A of https://arxiv.org/pdf/2011.03494

<img width="1248" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/7c412f56-8ca6-465f-beef-09b15427101d">


The unary iteration base class supports specifying multiple selection registers. You can look at implementation of `ApplyGateToLthQubit` or `SelectedMajoranaFermion` gates; both of which supports multiple selection registers

Reply:
> Consider making `ops` a `Union[HasLength, Tuple[Bloq, ...]]` to support symbolics.

One challenge with making the list symbolic and length- or shape-only here is that it's not clear how to compute the call graph if we don't know concretely what the constituent bloqs are --- the number of them doesn't seem to be enough information. We had the same dilemma with `TensorProduct` in #1094.

Reply:
> consider accepting a ndarray instead of a Tuple and make it a `Union[Shaped, NDArray[Bloq]]` so that users can specify N-dimensional SELECT operations.

Setting aside the symbolic issue of `HasLength`/`Shaped` above, I took a stab at switching to `NDArray`. The issue is that `NDArray` is a mutable/non-hashable type, which breaks the use of bloqs as keys in sets/dicts by the call graph infrastructure:

```
qualtran/_infra/bloq.py:343: in call_graph
    return get_bloq_call_graph(self, generalizer=generalizer, keep=keep, max_depth=max_depth)
qualtran/resource_counting/_call_graph.py:241: in get_bloq_call_graph
    _build_call_graph(bloq, generalizer, ssa, keep, max_depth, g=g, depth=0)
qualtran/resource_counting/_call_graph.py:142: in _build_call_graph
    g.add_node(bloq)
../.virtualenvs/qualtran/lib/python3.11/site-packages/networkx/classes/digraph.py:460: in add_node
    if node_for_adding not in self._succ:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = ApplyLthBloq(ops=array([TGate(is_adjoint=False), Hadamard(), ZGate(), XGate()],
      dtype=object), selection_regs=None, control_val=1)

    def __hash__(self):
>           return hash((
                    -8356619234005614914,
                    self.ops,
                    self.selection_regs,
                    self.control_val,
                ))
E           TypeError: unhashable type: 'numpy.ndarray'
```

I haven't found examples of existing bloqs that have `NDArray` attributes and overcome this problem. The closest examples are bloqs that make the attribute immutable by fixing its dimension, e.g. to `Tuple[Tuple[Bloq, ...], ...]`. But that isn't what we want either.

Reply:
One way to potentially avoid the above issue is to emulate `ApplyGateToLthQubit` and use a `Callable[..., Bloq]` rather than a `NDArray[Bloq]` to define the nth bloq. Assuming the callable is hashed via its pointer this may work, but I haven't worked out how good the resulting interface looks to the bloq user.

It seems `SelectedMajoranaFermion` does not encounter this issue because it only stores a single gate that it applies selectively.

Reply:
Here is an example - https://github.com/quantumlib/Qualtran/blob/3315602fc762b3b94300d449aebe2a23fbddb29a/qualtran/bloqs/data_loading/qrom_base.py#L153

You define the attribute as an `attrs.field()` and use a custom function to convert an NDArray to a flattened tuple that can be used in comparisons and hashing by overriding the `eq=_ndarray_to_tuple`

Reply:
I took a stab at using the `Callable` approach in the latest commit, but I'm not sure the code is more robust than in the one-dimensional case. Happy to receive feedback.

Reply:
Sorry, what's the benefit of using a `Callable` instead of directly expecting an NDArray in the input and using the approach I described to make the NDArray hashable / comparable? A callable has a drawback that its not (a) comparable (b) hashable and (c) serializable. Expecting an NDArray directly would fix all these issues.

Reply:
Thanks for the suggestion of the conversion. I'll give it a try to replace the `Callable`.